src/utils/authMiddlewares.js

```diff
@@ -79,6 +80,58 @@ const authenticationMiddlewareExpress = async (app) => {
   });
 };
 
+const longTimeoutEndpoints = [{ urlMatcher: /^\/v1\/experiments\/.{32}\/cellSets$/, method: 'PATCH' }];
+
+const runningOnLocalhost = (req) => {
+  const ip = req.connection.remoteAddress;
+  const host = req.get('host');
+
+  return ip === '127.0.0.1' || ip === '::ffff:127.0.0.1' || ip === '::1' || host.indexOf('localhost') !== -1;
+};
+
+const runningInsideCluster = async (req) => {
+  const insideCluster = await Promise.promisify(dns.reverse(req.ip));
+
+  return insideCluster;
+};
+
+const checkAuthExpiredMiddleware = async (req, res, next) => {
+  if (!req.user) {
+    next();
+    return;
+  }
+
+  const expirationDate = req.user.exp * 1000;
+
+  const timeLeft = expirationDate - Date.now();
+
+  if (timeLeft > 0) {
+    next();
+    return;
+  }
+
+  const pointingAtValidEndpoint = longTimeoutEndpoints
+    .some(({ urlMatcher, method }) => req.method === method && urlMatcher.test(req.url));
+
+  const hour = 1000 * 60 * 60;
+  const sixHours = 6 * hour;
+  const overranLongExpiration = timeLeft < -sixHours;
+
+  if (!pointingAtValidEndpoint || runningOnLocalhost(req) || overranLongExpiration) {
+    next(new UnauthenticatedError('token has expired'));
+    return;
+  }
+
+  // This operation ran apart from the rest because it takes longer
+  const isRunningInsideCluster = await runningInsideCluster(req);
+  if (isRunningInsideCluster) {
```

Comment:
simialrly here


```
  // This operation ran apart from the rest because it takes longer
  const isRunningInsideCluster = await runningInsideCluster(req);

  // if request originated from outside localhost/cluster 
  if (!isRunningInsideCluster && !runningOnLocalhost(req)) {
```